### DISCUSSION TITLE
Need to install Go to use Hugo modules

### DISCUSSION THREAD

**Discussion by iinfin** [vercel/community#38] [Help] [answered]
I'm trying to use [Hugo modules](https://gohugo.io/hugo-modules/use-modules/) with a theme. And I'm getting an error saying Go needs to be installed.

```
Error: we found a go.mod file in your project, but you need to install Go to use it.
See https://golang.org/dl/.: module "github.com/u29dc/hbp" not found;
either add it as a Hugo Module or store it in "/vercel/79b68035/themes".: module does not exist
```

[Go runtime documentation](https://vercel.com/docs/runtimes#official-runtimes/go/go-dependencies) says `go.mod` file is automatically detected to use with Go. Isn't it supposed to be already installed?
Am I missing something, was there an extra configuration I needed to set?
Couldn't see it in issues or discussions either.

**Comment by styfle** [comment, marked as answer]
Hi @iinfin 

The Go documentation is for Go Serverless functions.

We don't provide Go at build time for static builds like Hugo, you will need to install Go manually if you need it.

We have an internal Story 6785 tracking this feature request to preinstall Go for all builds, thanks!

**Reply by iinfin** [reply]
Alright then, thanks!

I'm not sure what the implications/disadvantages would be but yeah I feel like they should always be preinstalled.
I hope that's feasible, looking forward to it!

**Reply by rchamarthy** [reply]
@styfle Any update on this feature? If there is a place that we could contribute, I will be happy to take a stab at it.

**Reply by berot3** [reply]
Haven't found any more infos about this. Is this still something that is being worked on? I assume Vercel will not support modules anytime soon.

**Reply by ainsleyclark** [reply]
Hi. 
Just wondering if this is still in the pipeline? If it's not would you be able to recommend the best way to install go at build time on vercel?
Thanks

**Reply by suzannealdrich** [reply]
@ainsleyclark - Here's how I made it work. Override the Build Command in your project's settings, and add `golang` for yourself.
`amazon-linux-extras install golang1.11 && ls && echo "=== build it ===" && hugo --gc --minify --ignoreCache --verbose`
<img width="1293" alt="Screen Shot 2023-02-27 at 8 59 03 PM" src="https://user-images.githubusercontent.com/1848316/221758394-7fca0771-cb40-4633-b890-33ad5330e602.png">

**Reply by ainsleyclark** [reply]
Thanks @suzannealdrich !

Do you know if 1.19 is supported? 1.11 seems quite old.

Many thanks.

**Reply by suzannealdrich** [reply]
I tried a bunch of more recent versions, but sadly I think that's the latest you can get. 😞

**Reply by barelyhuman** [reply]
@ainsleyclark The latest answer below should help

**Comment by barelyhuman** [comment]
Probably already found the following solution online, but I'm going to write it down for others. 

If you need go , you can use the command `amazon-linux-extras install golang1.11` in your build step and it'd install go but I'm sure hugo needs 1.16 and you are better off using a go binary package for something like this. 

so you can instead do this

1. Write a bash script to curl the latest releases of Hugo and download the linux variant of the binary and run it during the buildstep 
2. Run the binary during the build 

There used to be gobinaries.com, which would work but the app can't be used for builds over requirement of go lang 1.15

**Reply by barelyhuman** [reply]
Update: 
https://goblin.reaper.im/
A fork of gobinaries with version on 1.16, should work for you. 

As for 1.17, I'm shorthanded but I'll see if I can get that working soon

**Reply by suzannealdrich** [reply]
I tried this but for some reason had problems deploying, with following error:
`Error: No Output Directory named "public" found after the Build completed. You can configure the Output Directory in your Project Settings. Learn More: https://vercel.link/missing-public-directory`

Resolved with following in my overridden build step:
`amazon-linux-extras install golang1.11 && hugo --gc --minify --ignoreCache --verbose`

**Reply by barelyhuman** [reply]
So, I just tried building `hugo` locally the same way goblin does it and the time taken for building is longer than the network can handle, 
I'm still figuring out a solution for such build times.

**I'd discard this as a solution till then**

Thanks @suzannealdrich for bringing this up

**Reply by barelyhuman** [reply]
If the amz linux works it's cool, Goblin won't be able to wait for 10 mins to build this from scratch on request so it's just easier to use something like https://bina.egoist.sh/
or the above solution

**Reply by barelyhuman** [reply]
The link for the above has changed to 
https://bina.egoist.dev/ 
for anyone who'd still like to use prebuilt binaries instead

**Comment by imfing** [comment]
Solution for future searchers:

All you need to do is setting the `Install Command` to `yum install golang`:

![image](https://github.com/vercel/community/assets/5097752/b5e3f2a1-ca0f-421e-96b1-978b679361d6)

The golang available in the aforementioned `amazon-linux-extras` is really outdated.

Reference: [Installing something else in the build image | Build image | Vercel Docs](https://vercel.com/docs/deployments/build-image#installing-something-else-in-the-build-image)
